Sources/SecureXPC/Client/XPCClient.swift

```diff
@@ -158,42 +130,93 @@ public class XPCClient {
+    ///   - ofType: There are multiple different types of XPC clients and normally you do not need to concern yourself with this. However, if you are trying to
+    ///             create a client for an XPC Mach service *and* you have an XPC service with a CFBundleIdentifier value that's the same as the name of
+    ///             that Mach service, then you must call this function and explicitly set the type to ``ServiceType/machService``. This is because auto
+    ///             detection will always choose the XPC service if one exists with the provided name.
     /// - Returns: A client configured to communicate with the named service.
-    public static func forMachService(named machServiceName: String) -> XPCClient {
-        XPCMachClient(machServiceName: machServiceName)
+    public static func forService(named serviceName: String, ofType type: ServiceType = .autoDetect) -> XPCClient {
+        switch type {
+            case .autoDetect:
+                if bundledXPCServiceIdentifiers.contains(serviceName) {
+                    return XPCServiceClient(xpcServiceName: serviceName)
+                } else {
+                    return XPCMachClient(machServiceName: serviceName)
+                }
+            case .xpcService:
+                guard bundledXPCServiceIdentifiers.contains(serviceName) else {
+                    let message = "There is no bundled XPC service with name \(serviceName)\n" +
+                                  "Available XPC service names are:\n" +
+                                  bundledXPCServiceIdentifiers.joined(separator: "\n")
+                    fatalError(message)
+                }
+                
+                return XPCServiceClient(xpcServiceName: serviceName)
+            case .machService:
+                return XPCMachClient(machServiceName: serviceName)
+        }
     }
+    
+    /// The `CFBundleIdentifier` values for every `.xpc` file in the `Contents/XPCServices` of this app (if it exists).
+    private static let bundledXPCServiceIdentifiers: Set<String> = {
+        var xpcServiceBundleIdentifiers = Set<String>()
+        let servicesDir = Bundle.main.bundleURL.appendingPathComponent("Contents")
+                                               .appendingPathComponent("XPCServices")
+        if let bundleNames = try? FileManager.default.contentsOfDirectory(atPath: servicesDir.path) {
+            for bundleName in bundleNames {
+                if bundleName.hasSuffix(".xpc"),
+                   let bundle = Bundle(url: servicesDir.appendingPathComponent(bundleName)),
+                   let bundleIdentifier = bundle.infoDictionary?[kCFBundleIdentifierKey as String] as? String {
+                    xpcServiceBundleIdentifiers.insert(bundleIdentifier)
+                }
+            }
+        }
+        
+        return xpcServiceBundleIdentifiers
+    }()
```

Comment:
Also, this code (and my different approach to it) ignore the IDs of all bundles that:

1. Don't end with `xpc`
2. Aren't a valid loadable `Bundle`.
3. Have a `nil` `infoDictionary`
4. Have an `infoDictionary` with no value for `kCFBundleIdentifierKey`
5. Have a non-String value for `kCFBundleIdentifierKey`.

Should we be a bit more strict about this?

Perhaps something along the lines of 

```suggestion
    /// The `CFBundleIdentifier` values for every `.xpc` file in the `Contents/XPCServices` of this app (if it exists).
    private static let bundledXPCServiceIdentifiers: Set<String> = {
        let servicesDir = Bundle.main.bundleURL.appendingPathComponent("Contents")
                                               .appendingPathComponent("XPCServices")

        guard let bundleNames = try? FileManager.default.contentsOfDirectory(atPath: servicesDir.path) else {
            return []
        }

        let xpcBundleNames = bundleNames.compactMap { bundleName -> String? in
            guard bundleName.hasSuffix(".xpc") else { return nil }

            guard let bundle = Bundle(url: servicesDir.appendingPathComponent($0)) else {
                throw SomeError.missingBundle // When would this even be possible?
            }

            guard let infoDict = bundle.infoDictionary else {
                throw SomeError.missingInfoDictionary
            }

            guard let bundleID = infoDict[kCFBundleIdentifierKey as String] else {
                throw SomeError.missingBundleID
            }

            guard let bundleID as? String else {
                throw SomeError.malformedBundleID
            }

            return bundleID
        }

        return Set(xpcBundleIDs)
    }()
```

Reply:
Happy to switch to the `compactMap` & `filter` approach. I have to imagine it's a bit less efficient due to the additional looping, but this must be _extremely_ marginal compared to the disk I/O this function involves to read the property list.

I don't think the latter approach is correct in the sense that there's nothing "wrong" with an API user having malformed XPC services that are otherwise unused. I think it'd be a rather odd if a side effect of trying to get an XPC client resulted in an error being thrown that a totally unrelated XPC service is for example lacking a bundle id.

Reply:
> I have to imagine it's a bit less efficient due to the additional looping

Also the optimizer can often detect the loop invariance and merge the two loops. I wouldn't worry about it in most cases, particularly if we don't expect the sequence to have less 10,000s of elements.

As [Eric Lippert puts it](https://ericlippert.com/2020/03/27/new-grad-vs-senior-dev/):

> Ha ha, nested for loops go brrrrrrrrrrrr…